### PR TITLE
fix: fix the bug in unmarshalling fhe type value

### PR DESCRIPTION
**Pull request by lovenoble**

### CODE REVIEW

File: fhevm/tee_interpreter.go [hunk 1]
```diff
@@ -405,57 +405,33 @@ func marshalTfheType(value any, typ tfhe.FheUintType) ([]byte, error) {
...
-		if len(data) < 8 {
-			return 0, fmt.Errorf("data too short for FheUint160")
-		}
 		return binary.BigEndian.Uint64(data[len(data)-8:]), nil
 	default:
 		// If it's a BigInt or unsupported, throw an error
 		return 0, fmt.Errorf("unsupported FheUintType: %s", typ)
 	}
 }
-
```
Comment (amaury1093) [inline fhevm/tee_interpreter.go:458]: add back this line break

File: fhevm/tee_interpreter.go [hunk 2]
```diff
@@ -405,57 +405,33 @@ func marshalTfheType(value any, typ tfhe.FheUintType) ([]byte, error) {
 func unmarshalTfheType(data []byte, typ tfhe.FheUintType) (uint64, error) {
 	switch typ {
 	case tfhe.FheBool:
-		if len(data) < 1 {
-			return 0, fmt.Errorf("data too short for type %s", typ)
-		}
-		if data[0] == 0 {
+		if data[len(data)-1] == 0 {
```
Comment (amaury1093) [inline fhevm/tee_interpreter.go]: Not related this this line, but putting here:

in tee_cast.go:
```
	if castToType == tfhe.FheUint4 {
		value = value % 16
	}
```

These 3 lines should go inside `marshalTfheType` function

File: fhevm/tee_interpreter.go [hunk 3]
```diff
@@ -405,57 +405,33 @@ func marshalTfheType(value any, typ tfhe.FheUintType) ([]byte, error) {
 func unmarshalTfheType(data []byte, typ tfhe.FheUintType) (uint64, error) {
```
Comment (amaury1093) [inline fhevm/tee_interpreter.go]: @lovenoble This whole function could be removed, and replaced by one line:
```go
result := big.NewInt(0).SetBytes(data).Uint64()
```

That was the reason I wrote this comment:

https://github.com/Inco-fhevm/fhevm-go/blob/f0c6ba75e84bce4887145bd5a9fb9aa2728cb59f/fhevm/tee_interpreter.go#L45-L48

File: fhevm/tee_interpreter.go [hunk 4]
```diff
@@ -405,57 +405,33 @@ func marshalTfheType(value any, typ tfhe.FheUintType) ([]byte, error) {
...
-		if len(data) < 1 {
-			return 0, fmt.Errorf("data too short for type %s", typ)
-		}
-		return uint64(data[0] % 16), nil
-	case tfhe.FheUint8:
-		if len(data) < 1 {
-			return 0, fmt.Errorf("data too short for type %s", typ)
-		}
-		return uint64(data[0]), nil
+	case tfhe.FheUint4, tfhe.FheUint8:
```
Comment (amaury1093) [inline fhevm/tee_interpreter.go]: Also not related to this line, but in tee_cast.go, remove these 3 lines:

```go
	if ciphertext.FheUintType == castToType {
		return nil, errors.New("casting to same type is not supported")
	}
```

**Review by amaury1093 — CHANGES_REQUESTED** [review]
I think we should:
- either use this big switch case
- or use a one-liner `result := big.NewInt(0).SetBytes(data).Uint64()`

But we should definitely not mix up the two usages, it's getting confusing.

I also added some comments from the previous PR https://github.com/Inco-fhevm/fhevm-go/pull/24, that got merged even though I had pending review comments

File: fhevm/tee_interpreter.go (hunk 4 above)
Comment (lovenoble) [reply fhevm/tee_interpreter.go]: For fhe operations, casting to same type is not allowed. https://github.com/Inco-fhevm/fhevm-go/blob/main/fhevm/tfhe/tfhe_ciphertext.go#L2024
So this is just for compatibility.

Comment (amaury1093) [reply fhevm/tee_interpreter.go]: Ah, right, I missed this. Okay let's keep it.

File: fhevm/tee_interpreter.go (hunk 3 above)
Comment (lovenoble) [reply fhevm/tee_interpreter.go]: done

File: fhevm/tee_interpreter.go (hunk 2 above)
Comment (lovenoble) [reply fhevm/tee_interpreter.go]: done

File: fhevm/tee_interpreter.go (hunk 1 above)
Comment (lovenoble) [reply fhevm/tee_interpreter.go:458]: done

**Review by amaury1093 — APPROVED** [review]
LGTM

**Comment by amaury1093** [comment]
Could you fix the failing test before merging?

**Comment by lovenoble** [comment]
Yup. I had known about it. If we move the following lines
```go
	if castToType == tfhe.FheUint4 {
		value = value % 16
	}
```
into `marshalTfheType`, this causes the failures in test cases of other functions such as `rotr` for uint4 type. That's because `uint4` was considered the same way as `uint8` in some cases. So I was gonna raise another PR to fix them. Can you still merge this?

**Comment by amaury1093** [comment]
Code LGTM, but tests failing does seem worrying. Should be fixed by #29 as per @lovenoble. Exceptionally merging.

**Pull request merged**